include/fastrtps/utils/concurrent_queue.h

```diff
@@ -0,0 +1,139 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef _FASTRTPS_UTILS_CONCURRENT_QUEUE_H_
+#define _FASTRTPS_UTILS_CONCURRENT_QUEUE_H_
+
+#include <queue>
+#include <mutex>
+#include <condition_variable>
+
+namespace eprosima {
+namespace fastrtps {
+
+/**
+ * @brief ConcurrentQueue. A thread-safe shared queue.
+ *
+ * Multiple producers, multiple consumers thread safe queue.
+ * Since 'return by reference' is used this queue won't throw.
+ *
+ * Based on std::queue<T>.
+ */
+template<typename T, typename Sequence = std::deque<T>>
+class ConcurrentQueue
+{
+  using Queue = std::queue<T, Sequence>;
+
+public:
+
+  ConcurrentQueue()  = default;
+  ~ConcurrentQueue() = default;
+
+  ConcurrentQueue(const std::size_t count)
+    : queue_(typename Queue::container_type(count)) {}
+
+  // not-copyable
+  ConcurrentQueue& operator=(const ConcurrentQueue&) = delete;
+  ConcurrentQueue(const ConcurrentQueue& other)      = delete;
+
+  // not-movable
+  ConcurrentQueue& operator=(ConcurrentQueue&&)  = delete;
+  ConcurrentQueue(const ConcurrentQueue&& other) = delete;
+
+  /**
+   * @brief push. Move an object to the queue.
+   * @param item.
+   */
+  template <typename O>
+  void push(O&& item)
+  {
+//    static_assert(std::is_same<typename std::remove_cv<O>::type, T>::value,
+//                  "Unexpected argument type.");
+
+    std::unique_lock<std::mutex> lock(mutex_);
+    queue_.emplace(std::forward<O>(item));
+    lock.unlock();
+
+    has_data_.notify_one();
+  }
+
+  /**
+   * @brief try_pop. Return true immediately if successful retrieval.
+   * @param popped_item.
+   * @return true if an object was pop, false otherwise.
+   *
+   * @see wait_pop
+   */
+  bool try_pop(T& popped_item)
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    if (queue_.empty())
+      return false;
+
+    popped_item = std::move(queue_.front());
+    queue_.pop();
+    return true;
+  }
+
+  /**
+   * @brief wait_pop. Wait until an object is popped.
+   * @param popped_item.
+   *
+   * @see try_pop
+   */
+  T wait_pop()
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    while (queue_.empty())
+    {
+      has_data_.wait(lock);
+    }
+
+    auto popped_item = std::move(queue_.front());
+    queue_.pop();
+    return popped_item;
+  }
+
+  /**
+   * @brief empty.
+   * @return true if empty, false othewise.
+   */
+  bool empty() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return queue_.empty();
+  }
+
+  /**
+   * @brief size.
+   * @return size of the queue.
+   */
+  std::size_t size() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return queue_.size();
+  }
+
+protected :
+
+  std::condition_variable has_data_;
+  mutable std::mutex mutex_;
+
+  std::queue<T> queue_;
```

Comment:
Should this be `std::queue<T, Sequence> queue_;`?

Reply:
Good catch, this should be `Queue` defined at the top of the class.